### DISCUSSION TITLE
How to get flow.request.url for https site that does not exist?

### DISCUSSION THREAD

**Discussion by rain1** [mitmproxy/mitmproxy#5817] [Q&A] [answered]
Hi!

I have this simple addon that just prints requests to console.
```
class Counter:


    def __init__(self):
        print("I'm alive!")
    
    #@concurrent
    def request(self, flow):
        print("request url: " + flow.request.url)

    def response(self, flow):
        print("response request url: " + flow.request.url)


## TLS stuff


addons = [
    Counter()
]
```

Now when I visit site that does not exist: https://does.not.exist/some/imaginary/uri. 

When this site begins with HTTP then it works fine:
`request url: http://does.not.exist/some/imaginary/uri`

However when I visit same url with HTTP**S** then:
1) It complains: `error establishing server connection: [Errno 11001] getaddrinfo failed`
2) `request`method is skipped and `response` is used instead.
3) it prints: `response request url: does.not.exist:443` meaning that `flow.request.url` does not contain full url anymore.

Is it possible to make it so that when I visit this HTTP**S** `https://does.not.exist/some/imaginary/uri` url then I still get to access full url somehow?

I am working on caching proxy where all kinds of imaginable and unimaginable server errors are to be expected.  Currently I am in place where I have response for that request in my database but I can't serve it because mitmproxy strips the URI part of url and switching from HTTPS to HTTP is not an option for me. I have to get it work with HTTPS and for that I do need full url in `flow.request.url` variable, not just `does.not.exist:443` like it is now.

**Comment by mhils** [comment, marked as answer]
You can make this work by setting [`connection_strategy`](https://docs.mitmproxy.org/dev/concepts-options/#connection_strategy) to `lazy`. The downside is that mitmproxy cannot determine if the upstream server supports HTTP/2 during the handshake, and will offer HTTP/1 connections only.

**Reply by rain1** [reply]
Thanks!

Can this option also be set from code level? I have modified my mitmdump main method like that:
```
if __name__ == '__main__':
    sys.argv.append("-s")
    sys.argv.append("anatomy.py")
    sys.argv.append("--ssl-insecure")
    sys.argv.append("--no-http2")
    # ... more code
```

Can I add this like that here? Like `sys.argv.append("--options onnection_strategy=lazy")` or what would be the exact syntax?  Since I already append --no-http2 here then in my case I wouldn't loose anything anymore with that lazy connection_strategy? Also are there any web functionality that would break because of this restriction such as chats or streams or can they all be made to use HTTP1 with just some performance hit?

**Reply by mhils** [reply]
Yes, as per https://docs.mitmproxy.org/dev/concepts-options/#tools:
```python
sys.argv.append("--set")
sys.argv.append("connection_strategy=lazy")
```
> Also are there any web functionality that would break because of this restriction such as chats or streams or can they all be made to use HTTP1 with just some performance hit?

Generally speaking this should not be a problem. Of course there are theoretical edge cases, e.g. when the upstream server speaks HTTP/2 only and the HTTP/1 requests sent by the original client are incompatible. But it's nothing I would personally worry about.

**Comment by rain1** [comment]
Thanks!

Now it works. Btw just out of curiosity: why does it change the url like I described in my first post if this option is not set? As I understand technically you don't need server connection to be able to get full url from the request?
